Discord experimental

ajCameron opened the pull request:
So far can only send messages back to the channel the original message came from.

Will work on it soem more.

javajawa commented on .gitignore:
```diff
@@ -5,6 +5,14 @@
 /.local/
 /.cache/
 /venv/
+.idea
```
Set up your global gitignore up correctly! :3

javajawa commented on src/mewbot/ioconfigs/post_input/input.py:
```diff
@@ -0,0 +1,107 @@
+        self.logger.info("About to run web app")
+        # Re. The protected member access
+        # I'm not sure why the aiohttp people have made some of the design choices they have - perhaps this is not
+        # necessary and I've just misunderstood
```
https://github.com/javajawa/snerge-bot/blob/aiohttp/src/snerge/__main__.py#L64-L77

ajCameron replied:
Implemented.

javajawa commented on src/mewbot/ioconfigs/discord_io/input.py:
```diff
@@ -0,0 +1,102 @@
+            DiscordInputEvent,
+        }
+
+    async def run(self) -> None:
```
Does creating a task inside the run make sense? Doing so seems to make it harder for the caller to determine if something has fallen over, as it won't know what are the relevant tasks

ajCameron replied:
There was a reason I did it that way ... but I can't remember for the live of me what it was. So it probably wasn't great. Correcting.

ajCameron commented:
Think it's ready when you are - if you want to take a final look?

(Now, mypy is failing locally but passing remotely ... which, you know. Better than the other way round). :)

ajCameron commented:
Everything even still works ... somehow,

javajawa committed:
Update setup.py to new repo, using version 61

javajawa committed:
Update gitignore for IDEs etc.

javajawa committed:
Move demo code to a new 'example' folder

javajawa committed:
Update dependencies, switch oyaml to pyyaml

javajawa committed:
Remove some PyCharm linting errors from design docs

javajawa committed:
Refactor core logic into new module structure

- `mewbot.core` contains the base interface definitions
- `mewbot.component` contains the Component class and Registry for the auotmated YAML loading (the loading logic is in `mewbot.loader`)
- `mewbot.api.[version]` contains the base classes for that API version

All bot running logic depends on the interfaces from `mewbot.core`, rather than
any specific API implementation -- this means that the API modules can have a
different versioning cycle; so long as there are no interfaces changes in the
core there is no need for a major semantic version change.

Likewise, the mewbot.api modules do not know anything about the bot itself.
This can be packaged as a separate dependency, and then used to develop
component modules without having to pull in the complete bot codebase.

Commit:
Change Actions to run in async, and flesh out the async loop

Commit:
Add HTTP Post / Sockets IO logic and example code

Commit:
Add basic Discord IO and example

javajawa committed:
Fix linting issue

exachixkitsune commented on design-docs/README.md:
````diff
@@ -108,12 +108,14 @@ with the bot.
 configured). These configurations will then present a list of Inputs and Outputs
 which can be used in Triggers.
 
 ```python
````
Documentation Mismatch with Code : This does not appear to be a duplicate of any code anywhere in mewbot. Where should it be?

javajawa replied:
It's demonstrative rather than actual code. pycharm assumes that python code blocks should be correct and happens to lint them as well, so I fixed them up whilst doing the refactoring.

exachixkitsune commented on examples/trivial_discord_bot.py:
```diff
@@ -0,0 +1,148 @@
+        await self._queue.put(event)
+
+
+def load_secrets() -> Any:
```
Future Thoughts: Should this end up being a default/inbuilt thing to mewbot?

javajawa replied:
Secret storage in the configuration language (and the YAML backing store) is something I've not really got a solution for yet. Thoughts on that welcome in the design docs.

ajCameron replied:
Will also consider it! That's definitely a problem that has to be solved.

exachixkitsune commented on src/mewbot/behaviour.py:
```diff
@@ -75,29 +21,31 @@ def __init__(self) -> None:
 
-    def add(self, block: BehaviourComponent) -> None:
-        if not isinstance(block, BehaviourComponent):
+    def add(
```
Future Thoughts: One "setup" function which takes in the three Trigger / Condition / Action
Use case: setting up a behaviour on one line in bot init

javajawa replied:
Not needed, configuration should never be in python

javajawa replied:
(more likely is that this will get changed to adding an iterable of that the current type, a la
```python
add(self, *components: Union[ *Interface... ]
```
and then adding all of them by the existing method.

exachixkitsune commented on src/mewbot/behaviour.py:
```diff
@@ -75,29 +21,31 @@ def __init__(self) -> None:
-            self.conditions.append(block)
-        if isinstance(block, Action):
-            self.actions.append(block)
+        if isinstance(component, TriggerInterface):
```
Is there a potential use case for an induvidual component to be an instance of multiple of these? Can I have a component that is both a TriggerInterface and a ConditionInterface?
Do we want to accept those conditions, warn on those conditions, or terminate?

javajawa replied:
Doing so is explicity forbidden in the component API (https://github.com/ajCameron/mewbot/blob/2990b63/src/mewbot/component.py#L43-L46) but nothing prevents alternative implementations of these interfaces doubling up.

exachixkitsune commented on src/mewbot/behaviour.py:
```diff
@@ -2,70 +2,16 @@
-        Condition
-    ]  # conditions on the action being called when it is matched (e.g. "is subscribed")
-    actions: List[Action]  # the things to do if an event matches, and the conditions are met
+    triggers: List[TriggerInterface]
```
Minor/Readibility: Comments on each line were lost

javajawa replied:
Intentional as inline comments should not be used for documentation (probably a hold over from copy pasting from the design documents?)
Currently pretty much everything is missing both doucmentation and tests

exachixkitsune commented on examples/trivial_discord_bot.py:
```diff
@@ -0,0 +1,148 @@
+    print_behavior.add(DiscordPrintAndReplyAction())
+
+    # - IOConfig
+    post_ioconfig = PostIOConfig()
```
Future Thoughts/Shortcuts:If PostIOConfig is required for configuring Discord, do we want to make a system that automatically does that pipeline?
Or do we only ever do one add_io_config to bot, and so post_ioconfig should contain everything?

javajawa replied:
Neither, really.
Building on both https://github.com/mewler/mewbot/tree/2990b637184105bdba773f0cff50007f968632f7/design-docs#input and the principle of configuration of the bot not being driven by python code, the solution is that the `mewbot.loader` will be able to load IOConfig objects of arbitary classes as needed, and those classes will automatically output the Input/Outputs that they support. (The method of building it up here is essentially the wrong way around)

javajawa replied:
https://github.com/mewler/mewbot/pull/4/commits/c8c08db127b13daa3d56f1c157a72cbbc2468fe4 is a good example of how it is envisioned to work!

exachixkitsune commented on src/mewbot/http_post.py:
```diff
@@ -0,0 +1,117 @@
+    def __init__(self) -> None:
+        super().__init__()
+
+        self.inputs = []
```
Minor/Missing:These have gets but not sets/adds; https://github.com/mewler/mewbot/blob/fe092a331db784ebb88f04028136a919a3f1656c/examples/trivial_discord_bot.py#L141 adds to them.

javajawa replied:
I think I have fixed this!

exachixkitsune commented on src/mewbot/bot.py:
```diff
@@ -4,28 +4,38 @@
     datastores: Dict[str, DataSource[Any]]  # Data sources and stores for this bot
 
+    def __init__(self) -> None:
+        self.name = "not_set"
```
Minor/Missing:This has no inherent gets or sets.

exachixkitsune commented on examples/trivial_discord_bot.py:
```diff
@@ -0,0 +1,148 @@
+# Remember when wiring random services together to see what you could make happen?
```
Overall Q for trivial_discord_bot.py; is this the eventual workflow we're going for with mewbot, or is this a temporary example for dev?

ajCameron replied:
My aim when putting it together was very tempoary example for dev. Just wanted something with a working event loop to play around with/preform basic validation of class drafts.

"example" is a bad name for this concept. In future, will tag such things as "scratch_{name}.py"

exachixkitsune commented on setup.py:
```diff
@@ -11,14 +11,14 @@
-    author_email="[Ben public email for this project] & mewbot@quicksilver.london",
+    author="Benedict Harcourt & Alex Cameron",
+    author_email="mewbot@tea-cats.co.uk & mewbot@quicksilver.london",
     description="Light weight text based generic irc framework",
```
Should be "YAML-driven Bot Framework", or something similar?

ajCameron replied:
I definitely agree. Will update.

exachixkitsune commented on src/mewbot/demo.py:
```diff
@@ -0,0 +1,32 @@
+
+    def __init__(self) -> None:
+        if not hasattr(self, "_channel"):
+            self._channel = "oops"
```
I know it's only a demo; should this error or be otherwise a warning; or does this transit to a warning/error later because "oops" is generally not a channel?

javajawa replied:
This is in general an interesting question because of the order of operations.
With the way the metaclass is setup, `ComponentRegistry.__call__` invokes `Component.__new__`, then sets all settable property decorators, then calls `Component.__init__` with any parameters not previously used. Mostly it is done in this order so that calls to create an object don't rely toooo much on the class's `__init__`, and also that that `__init__` may assume that all the properties are in place.

However, at the same time, because python classes are ducktyped, if a property is not set in the `__new__`, we still need a way to have the `__init__` create the underlying property if it was not set (or error out).

Potential options for that are:
 - Forcing all properties to be required in the `__new__` call (i.e. you can not not specify the properties) and thus ensuring the property is always defined
 - Default value on the property in the class definition, which will get overwritten by any call from `__new__`
 - This kind of checked operation in current code
 - Change how the loader works that setup happens after `__init__`

javajawa replied:
I have solved this specific example by just not having an `__init__` function, and a field default of `""`

exachixkitsune commented on src/mewbot/bot.py:
```diff
@@ -4,28 +4,38 @@
     datastores: Dict[str, DataSource[Any]]  # Data sources and stores for this bot
 
+    def __init__(self) -> None:
+        self.name = "not_set"
```
Apparently unused unless I cannot find where it's used.

javajawa replied:
Field is included as per spec (https://github.com/mewler/mewbot/tree/mainline/design-docs#bot) but doesn't actually really do anything -- at least until building out the UI and API structures starts happening.

exachixkitsune commented on src/mewbot/bot.py:
```diff
@@ -68,46 +84,75 @@ class BotRunner:
+        self.logger.info("Starting main event loop")
+
+        # Startup the inputs
+        self.logger.info("self.inputs: %s", ",".join(map(str, self.inputs)))
```
Not checked, but want to note a moment of care may need to happen here if the stringification of self.inputs captures the configuration of each input and any auth tokens if they are bound up in that. (which then gets logged presumably...)

javajawa replied:
By the spec, they should not just be dumping dicts of their own state (https://github.com/mewler/mewbot/tree/mainline/design-docs#human-readable-descriptions), but the burden of defining sensible str representation of components is up to the people implementing that component.

exachixkitsune commented on src/mewbot/loader.py:
```diff
@@ -0,0 +1,44 @@
+def load_component_from_yaml(config: ConfigBlock) -> Component:
+    required_keys = ConfigBlock.__annotations__.keys()  # pylint: disable=no-member
+
+    if set(config.keys()) != set(required_keys):
```
Should this be != or should it be A contains B?

javajawa replied:
Great question!
I left this as a direct equality check because it means that extra incoming properties will cause an error, making schema mistakes easier to detect.

There's a reasonable chance that there will be multiple TypedDicts representing different complex structures (such as a behaviour having triggers/conditions/actions) which is going to make this section of code more complex.

javajawa replied:
This has been changed to require a superset, as that was necessary for Behaviour loading.

exachixkitsune reviewed (dismissed):
oops accidently added a bunch of comments before someone let me know that was wrong.
I've resolved any "conversations" which looked like my query had been answered.
No major comments, only minors. And, let's be honest, many of those are possibly me pointing out things that are like that for a reason.

javajawa committed:
Convert 'Behaviour' to be an Interface + API

 - Behaviour class moved from mewbot.behaviour to mewbot.api.v1
 - Some checks for API Versions have been disabled

javajawa committed:
Refactor running loop in BotRunner

 - Properties are now marked as 'protected'
 - The bots name is now passed to the constructor.
 - Input/Output queue tasks now check on a timeout, so they can notice the shutdown command
 - Input runners are now cancelled when the loop ends
 - Generally cleaned up the event loop handling

javajawa committed:
Add support to load Behaviours and Bots from YAML

javajawa committed:
Overall socket example to run from YAML

javajawa committed:
Rework all the 'trivial' examples to be YAML driven

 - IO code bases have been moved to a mewbot.io module.
 - Examples are now in YAML with a single runner.
 - Discord and HTTP IOs have been rewritten as simpler configs.

javajawa committed:
Update types-PyYAML requirement

ajCameron committed:
mypy linting fixes - hopefully

ajCameron committed:
black does not like whitespace at file top

Cannot see a better solution for the mypy errors than just ignoring them

ajCameron committed:
Updated short description

To "Lightweight, YAML-driven, text based, generic irc Bot framework"

ajCameron commented:
Leaving in existence incase squash was the wrong way to go
